projects/storefrontlib/src/shared/components/media/media.service.ts

```diff
@@ -1,108 +1,126 @@
-import { BreakpointService } from '../../../layout/breakpoint/breakpoint.service';
-import { BREAKPOINT } from '../../../layout/config/layout-config';
-import { Media, MediaFormats } from './media.model';
-
-/** the default format is used for browsers that do not support   */
-const DEFAULT_MEDIA_FORMAT = 'tablet';
+import { Image, OccConfig } from '@spartacus/core';
+import { MediaConfig } from './media.config';
+import { Media, MediaContainer, MediaFormatSize } from './media.model';
 
+/**
+ * Service which generates media URLs. It leverage the MediaContainer and MediaFormats so
+ * that URLs and sizes are generated for the same media. This helps to improve performance
+ * across difference devices and layouts.
+ *
+ * Media formats are optional, but highly recommended. The format will help the browser to
+ * identify the right media for the right experience.
+ *
+ * The MediaService will generate absolute URLs in case relative URLs are provided for the Media.
+ * The baseUrl is read from the `occConfig.backend.media.baseUrl` or
+ * `occConfig.backend.occ.baseUrl`.
+ */
 @Injectable({
   providedIn: 'root',
 })
 export class MediaService {
-  constructor(
-    protected config: OccConfig,
-    protected breakpointService: BreakpointService
-  ) {}
+  /**
+   * The media formats sorted by size. The media format representing the smallest
+   * size is sorted on top.
+   */
+  protected sortedFormats: { code: string; size: MediaFormatSize }[];
+  protected reversedFormats: { code: string; size: MediaFormatSize }[];
 
-  private get mediaFormats(): MediaFormats[] {
-    return [
-      {
-        code: 'mobile',
-        threshold: this.breakpointService.getSize(BREAKPOINT.xs),
-      },
-      {
-        code: 'tablet',
-        threshold: this.breakpointService.getSize(BREAKPOINT.sm),
-      },
-      {
-        code: 'desktop',
-        threshold: this.breakpointService.getSize(BREAKPOINT.md),
-      },
-      {
-        code: 'widescreen',
-        threshold: this.breakpointService.getSize(BREAKPOINT.lg),
-      },
-    ];
+  constructor(
+    protected occConfig: OccConfig,
+    protected mediaConfig: MediaConfig
+  ) {
+    this.sortFormats();
   }
 
-  getMedia(container, format?: string, alt?: string): Media {
+  /**
+   * Returns a `Media` object with the main media (`src`) and various media (`src`)
+   * for specific formats.
+   */
+  getMedia(
+    mediaContainer: MediaContainer | Image,
+    format?: string,
+    alt?: string
+  ): Media {
+    if (!mediaContainer) {
+      return;
+    }
+
+    const mainMedia: Image = mediaContainer.url
+      ? mediaContainer
+      : this.resolveMedia(mediaContainer as MediaContainer, format);
+
     return {
-      src: this.getMainImage(container, format),
-      srcset: this.getSrcSet(container),
-      alt: alt || this.getAlt(container, format),
+      src: this.resolveAbsoluteUrl(mainMedia?.url),
+      alt: alt || mainMedia?.altText,
+      srcset: this.resolveSrcSet(mediaContainer),
```

Comment:
It seems that specifying `srcset` without `sizes` works as implicit size 100%
We might consider adding `sizes` here (and in the media component)

Reply:
I like to postpone the introduction of sizes. The main issue here is the relation with sizes and viewport, where as images are often used inside a smaller layout; the viewport size would not really contribute here. 

I've "fixed" this (well, ensure it's aligned with the current approach), by introducing a CSS rule (`object-fit: scale-down`) that guaranties that the product image will not be rendered larger than the actual image.

Reply:
👍